Changed API to support Promises

mikepie1 opened the pull request:
Addresses issue #38. Left onEvent events as callbacks but changed others to Promises

Commit:
Changed API to support Promises

Addresses #38. Left onEvent events as callbacks but changed others to
Promises

frivoal commented:
Thanks for getting promises in.

However, I am still unsure about not keeping callbacks as well. I assume that since this is coming from you, Edge expects to support this API with promises, so no Polyfill will be needed in that direction.

However, do you also expect to not support callbacks? How about everyone else? whether it is implemented in C++ or Javascript ultimately matters very little. If browsers ship with both, authors will use either, and published extensions will depend on both, and new browser vendors wanting to support existing content will have to implement both.

As I suspect a polifyl to support chrome will need to do more than just `browser=chrome` and will have more differences to paper over, it's ok if we don't spec callbacks if for chrome compatibility.

But if browsers that support expensions natively under the `browser` namespace do take callbacks, I think they should stay in the spec.

Commit:
More details about localization, i18n, messages.json

Addresses issues #23 and #24 regarding codes, matching, tokens,
placeholders and predefined messages.

frivoal commented:
So we've had a discussion about this, and concluded we couldn't conclude yet. Both Microsoft and Mozilla intent to focus on promises rather than callbacks, but the timeline isn't fully clear, and whether the implementations will be promises only in the long run or whether they will keep callbacks as well for compatibility reasons is not yet fully decided.

In that light, it makes sense for now to specify promises only, and revisit in a while. If support for callbacks is staying in implementations, it will be worth documenting them as well (possibly with copious warnings about them being the legacy deprecated approach).

Commit:
Corrected ResSpec 1 error and 566 ReSpec warnings. The warnings were due to missing <dfn> definitions. I added placeholder <dfn> entries for all APIs("TBD_define_here"), and I created a new issue to track the revisions that will be needed to address this. Also, ReSpec kept generating a warning for the following 9 APIs (and I'm not sure why), so I added "_TBD" as a suffix to these and the warning went away. Need to figure out why this happens and correct it for ReSpec. APIs that generated warnings are: contextMenus.onClicked; i18n.getMessage; pageAction.onClicked; pageAction.setIcon; pageAction.setPopup; pageAction.getPopup; tabs.sendMessage; webRequest.onCompleted; windows.getCurrent
